frontend/src/app/main/data/workspace/thumbnails.cljs

```diff
@@ -0,0 +1,174 @@
+;; This Source Code Form is subject to the terms of the Mozilla Public
+;; License, v. 2.0. If a copy of the MPL was not distributed with this
+;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
+;;
+;; Copyright (c) UXBOX Labs SL
+
+(ns app.main.data.workspace.thumbnails
+  (:require
+   [app.common.data :as d]
+   [app.common.uuid :as uuid]
+   [app.main.data.workspace.changes :as dch]
+   [app.main.refs :as refs]
+   [app.main.repo :as rp]
+   [app.main.store :as st]
+   [beicon.core :as rx]
+   [potok.core :as ptk]))
+
+(defn force-render-stream
+  "Stream that will inform the frame-wrapper to mount into memory"
+  [id]
+  (->> st/stream
+       (rx/filter (ptk/type? ::force-render))
+       (rx/map deref)
+       (rx/filter #(= % id))
+       (rx/take 1)))
+
+(defn update-thumbnail
+  "Updates the thumbnail information for the given frame `id`"
+  [id data]
+  (let [lock (uuid/next)]
+    (ptk/reify ::update-thumbnail
+      IDeref
+      (-deref [_] {:id id :data data})
+      
+      ptk/UpdateEvent
+      (update [_ state]
+        (-> state
+            (assoc-in [:workspace-file :thumbnails id] data)
+            (cond-> (nil? (get-in state [::update-thumbnail-lock id]))
+              (assoc-in [::update-thumbnail-lock id] lock))))
+
+      ptk/WatchEvent
+      (watch [_ state stream]
+        (when (= lock (get-in state [::update-thumbnail-lock id]))
+          (let [stopper (->> stream (rx/filter (ptk/type? :app.main.data.workspace/finalize)))
+                params {:file-id (:current-file-id state)
+                        :object-id id}]
+            ;; Sends the first event and debounce the rest. Will only make one update once
+            ;; the 2 second debounce is finished
+            (rx/merge
+             (->> stream
+                  (rx/take-until stopper)
+                  (rx/filter (ptk/type? ::update-thumbnail))
+                  (rx/filter #(= id (:id (deref %))))
+                  (rx/debounce 2000)
+                  (rx/first)
+                  (rx/flat-map
+                   (fn [event]
+                     (let [data (:data @event)]
```

Comment:
since event is not used directly, you can remove the let just adding a new maps

``
(rx/map deref)
(rx/map :data)
```

Reply:
ok